skymp5-server/cpp/server_guest_lib/database_drivers/FileDatabase.cpp

```diff
@@ -20,16 +20,30 @@ FileDatabase::FileDatabase(std::string directory_,
 
 size_t FileDatabase::Upsert(const std::vector<MpChangeForm>& changeForms)
 {
-  auto p = pImpl->changeFormsDirectory;
+  std::filesystem::path p = pImpl->changeFormsDirectory;
   size_t nUpserted = 0;
 
   for (auto& changeForm : changeForms) {
     std::string fileName = changeForm.formDesc.ToString('_') + ".json";
-    auto filePath = p / fileName;
-    std::ofstream f(filePath);
+    auto tempFilePath = p / (fileName + ".tmp"), filePath = p / fileName;
+
+    std::ofstream f(tempFilePath);
     if (f) {
       f << MpChangeForm::ToJson(changeForm).dump(2);
     }
+
+    if (!f.fail()) {
+      f.close(); // otherwise rename fails on Windows
```

Comment:
I think that this comment doesn't make any sense anyway. In any case, we need to make sure the file was flushed first